### PR TITLE
Allow lazy init of GPU pointer, so even if compiled with GPU support, won't access GPU unless run GPU algorithm like gpu_hist.

### PR DESCRIPTION
**Pull request by pseudotensor**

### CODE REVIEW

**Comment by pseudotensor** [comment]
@RAMitchell

**Comment by RAMitchell** [comment]
@canonizer please review.

File: src/common/host_device_vector.cu [hunk 1]
```diff
@@ -13,7 +13,7 @@ struct HostDeviceVectorImpl {
     : device_(device), on_d_(device >= 0) {
     if (on_d_) {
       dh::safe_cuda(cudaSetDevice(device_));
-      data_d_.resize(size, v);
+      data_d_->resize(size, v);
```
Comment (canonizer) [inline src/common/host_device_vector.cu]: `data_d_` is not initialized here.

File: src/common/host_device_vector.cu [hunk 2]
```diff
@@ -24,8 +24,9 @@ struct HostDeviceVectorImpl {
     : device_(device), on_d_(device >= 0) {
     if (on_d_) {
       dh::safe_cuda(cudaSetDevice(device_));
-      data_d_.resize(init.size());
-      thrust::copy(init.begin(), init.end(), data_d_.begin());
+      if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
```
Comment (canonizer) [inline src/common/host_device_vector.cu]: `data_d_` is not initialized here, so no need to check it before calling `reset()`.

File: src/common/host_device_vector.cu [hunk 3]
```diff
@@ -35,13 +36,21 @@ struct HostDeviceVectorImpl {
   void operator=(const HostDeviceVectorImpl<T>&) = delete;
   void operator=(HostDeviceVectorImpl<T>&&) = delete;
 
-  size_t Size() const { return on_d_ ? data_d_.size() : data_h_.size(); }
+  size_t Size() {
+     if (on_d_) {
+       if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
```
Comment (canonizer) [inline src/common/host_device_vector.cu]: Maybe just assume that `data_d_` is initialized if `on_d_` is true. In this way, many unnecessary checks can be avoided.

File: src/common/host_device_vector.cu [hunk 4]
```diff
@@ -35,13 +36,21 @@ struct HostDeviceVectorImpl {
...
+       return data_h_.size();
+     }
+  }
 
   int DeviceIdx() const { return device_; }
 
   T* DevicePointer(int device) {
     LazySyncDevice(device);
-    return data_d_.data().get();
+    if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
```
Comment (canonizer) [inline src/common/host_device_vector.cu]: `data_d_` should be initialized after `LazySyncDevice()`.

File: src/common/host_device_vector.cu [hunk 5]
```diff
@@ -87,17 +98,19 @@ struct HostDeviceVectorImpl {
       CHECK_EQ(device_, -1);
       device_ = device;
     }
-    if (data_d_.size() != this->Size()) {
+    if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
+    if (data_d_->size() != this->Size()) {
       dh::safe_cuda(cudaSetDevice(device_));
-      data_d_.resize(this->Size());
+      if (data_d_ == nullptr) data_d_.reset(new thrust::device_vector<T>());
```
Comment (canonizer) [inline src/common/host_device_vector.cu]: This check is redundant.

**Comment by RAMitchell** [comment]
Closing as should be solved in #3287

**Pull request closed without merging**